jerry-core/api/jerry.c

```diff
@@ -6428,6 +6429,140 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Value is not a Container or Iterator")));
+  }
+
+  ecma_collection_t *container_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t, ext_obj_p->u.cls.u3.value);
+  lit_magic_string_id_t lit_id = ext_obj_p->u.cls.u2.container_id;
+  uint32_t entry_count =  ECMA_CONTAINER_ENTRY_COUNT (container_p);
+  uint8_t entry_size = ecma_op_container_entry_size (lit_id);
+
+  uint32_t index = 0;
+  uint8_t iterator_kind = ECMA_ITERATOR__COUNT;
+  ecma_value_t *start_p = ECMA_CONTAINER_START (container_p);
+  bool is_value_iterator = false;
+
+  *is_key_value_p = false;
+
+  if (type == ECMA_OBJECT_TYPE_CLASS &&
+      (ext_obj_p->u.cls.type == ECMA_OBJECT_CLASS_MAP_ITERATOR ||
+        ext_obj_p->u.cls.type == ECMA_OBJECT_CLASS_SET_ITERATOR))
+  {
+    is_value_iterator = true;
+
+    ecma_value_t iterated_value = ext_obj_p->u.cls.u3.iterated_value;
+
+    if (ecma_is_value_empty (iterated_value))
+    {
+      return ecma_op_new_array_object_from_collection (collection_buffer, false);
+    }
+
+    ecma_extended_object_t *map_object_p = (ecma_extended_object_t *) (ecma_get_object_from_value (iterated_value));
+    lit_id = map_object_p->u.cls.u2.container_id;
+
+    container_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t, map_object_p->u.cls.u3.value);
+    entry_count = ECMA_CONTAINER_ENTRY_COUNT (container_p);
+    index = ext_obj_p->u.cls.u2.iterator_index;
+
+    if (index == entry_count)
+    {
+      return ecma_op_new_array_object_from_collection (collection_buffer, false);
+    }
+
+    entry_size = ecma_op_container_entry_size (lit_id);
+    start_p = ECMA_CONTAINER_START (container_p);
+
+    iterator_kind = ext_obj_p->u.cls.u1.iterator_kind;
+
+    if (iterator_kind == ECMA_ITERATOR_ENTRIES)
+    {
+      *is_key_value_p = true;
+    }
+  }
+  else if (lit_id == LIT_MAGIC_STRING_MAP_UL || lit_id == LIT_MAGIC_STRING_WEAKMAP_UL)
+  {
+    *is_key_value_p = true;
+  }
+
+  bool is_key_value_container = *is_key_value_p;
+
+  for (uint32_t i = index; i < entry_count; i+= entry_size)
+  {
+    ecma_value_t *entry_p = start_p + i;
+
+    if (ecma_is_value_empty (*entry_p))
+    {
+      continue;
+    }
+
+    if (is_value_iterator)
+    {
+      ext_obj_p->u.cls.u2.iterator_index = (uint16_t) (i + entry_size);
+
+      if (iterator_kind != ECMA_ITERATOR_ENTRIES)
+      {
+        ext_obj_p->u.cls.u2.iterator_index = (uint16_t) (i + entry_size);
+      }
+    }
+
+    if (is_key_value_container)
+    {
+      ecma_collection_push_back (collection_buffer, ecma_copy_value_if_not_object (*entry_p));
+      ecma_collection_push_back (collection_buffer, ecma_copy_value_if_not_object (entry_p[1]));
+    }
+    else if (iterator_kind == ECMA_ITERATOR_VALUES)
+    {
+      ecma_collection_push_back (collection_buffer, ecma_copy_value_if_not_object (entry_p[1]));
+    }
+    else
+    {
+      ecma_collection_push_back (collection_buffer, ecma_copy_value_if_not_object (*entry_p));
```

Comment:
Should we add an assert for the `iterator_kind` for expected kinds?

Reply:
I don't think there is a need for that. 

- An `ECMA_ITERATOR_ENTRIES` iterator will set the `is_key_value_container` as well as a Map or WeakMap (hence the `ECMA_ITERATOR__COUNT` doesn't matter there).
- The `ECMA_ITERATOR_VALUES` has it's own branch a`
- The `ECMA_ITERATOR_KEYS` will fall into the last one, as well as a Set or Weakset which is the only option where `ECMA_ITERATOR__COUNT` default should matter, but that container type is the same on keys-values in the jerry architecture if i understood it correctly.

Feel free to correct me if i'm wrong and/or missing the point in your comment.